### ISSUE TITLE
Wishlist: help for packages and classes needing hooks in \InputIfFileExists

### ISSUE THREAD

**Issue by callegar** [latex3/latex2e#225]
## Too many packages now relying on patching \InputIfFileExists for their functionality

A due premise. This is not a bug report, more of an enhancement request. And it is not really clear to me if it is acceptable at all. So please take my apology for the noise if it is not.

It is my feeling that there are now too many packages and classes needing to patch (or feeling the need to patch) `\InputIfFileExist` for their functionality. To mention a few:

- The KOMA script bundle has relied for ages on its own `scrlfile` package that patches `\InputIfFileExists` to provide hooks on file loading actions. All of KOMA relies on these hooks.

- The `filehook` package patches `\InputIfFileExists` to provide hooks on file loading actions. As of today, all the LuaLaTeX ecosystem indirectly rely on these hooks, via the `unicode-math` -> `lualatex-math` -> `filehook` dependency.  Specifically, `lualatex-math` needs file hooks to patch `amsmath` for lualatex at the right time.  Apparently, this is because `amsmath` is in code freeze and not doing any lualatex detection/adjustment itself.

- The `memoir` class patches `\InputIfFileExists` in its own way.

- The `paper` class patches `\InputIfFileExists` in its own way.

- The `fink` package patches `\InputIfFileExists` in its own way. Luckily, it is now considered obsolete in favor of `currfile` that relies on `filehook`.

- ...

This situation is very sub-optimal since all packages/classes try to patch the same LaTeX command in different ways so that breakage is almost sure when they are loaded together. Even if some packages like `filehook` try to implement *compatibility* helpers for other packages and classes (e.g., `filehook-scrlfile`, `filehook-memoir`, and soon some way in `scrlfile` to deal with `memoir`), the wide variety of packages and classes and versions thereof touching `\InputIfFileExist` make compatibility a hard combinatorial problem.

The issue has recently been amplified by three events:

1. The patching of an old issue in `\InputIfFileExists` in LaTeX2e (bug #109 )
2. The robustification of `\InputIfFileExists` in LaTeX2e
3. The increasing success of the LuaLaTeX engine

Specifically, 1 and 2 caused "udating waves" in all packages patching `\InputIfFileExists`. E.g.: LaTeX `\InputIfFileExists` changes; `filehook` is updated to match change in LaTeX; `scrlfile` is updated to match change in LaTeX; `filehook` needs to change again to keep compatibility to `scrlfile` that has changed; LaTeX changes again; `filehook` is updated once more, etc... At the same time, 3 makes `filehook` inherent in all lualatex flows. Together with 1 and 2 this meant that lualatex is often not in a usable state when coupled to KOMA script, memoir, paper, etc.

The large number of packages that feel in need to modify `\InputIfFileExists` to add hooks (justified or not this need may be), suggest that hooks in `\InputIfFileExists` could maybe be promoted to a core functionality.  IMHO, this could be done in two ways: either incorporating it in LaTeX directly or embracing a single solution (either `filehook` or `scrlfile`) as a latex base package, so that all other packages and classes are motivated to build their functionality on top of the standard mechanisms.  

Thanks for the attention and sorry for the long reading.

**Comment by FrankMittelbach** [comment]
I agree with you that there is the need to identify what kind of hooks/configuration points are need for different tasks and that those that are identified as "generally" sensible (suitably defined) should be promoted to be provided out of the box by the kernel to avoid the incompatible patching that had happened over time, because of the lack of them.

We are aware of this, and one of our goals is to provide this in the near future --- not only for `\InputIfFileExists` but also for other areas where the situation is similar. Alongside with this there has to be a hook management that provides better handling of how such hooks are being filled (and this way reducing/eliminating the need to juggle with package loading order to make things work).

What you could perhaps help with here is this:
 - provide a detailed analysis of what hooks are added by whom
 - document the conceptual purpose(s) of each hook is, e.g., what is hook A being used in KOMA for, etc.  

There are likely to be overlaps in the use cases, but probably with different solutions in different packages --- and clearly the answer here can't be that we add the combination of each an every hook. Instead  that we need to provide is a suitable minimal set that allows to resolve (preferably) all use cases without the need to patch.

So this kind of ground work would be quite helpful.

**Comment by callegar** [comment]
Thanks for your response and for the support to my enhancement request.

In the next days, I'll try to get in touch with the KOMA script and filehook authors to try to formalize a minimal set of functionalities that would be interesting to have. In fact, it is my feeling that what `memoir` and `paper` do could already be done using either of these to packages and if this has not happened yet, it is only because the authors of `memoir` and `paper` had not seen any of the two as *standard* (`scrlfile`) or stable (`filehook`, due to the "0.*" release number) to rely upon and they ended up rolling their own solutions. Package `fink` has been deprecated for a long time and I think that it is not very important at this point.

Hence the main people to involve are really the `filehook` and `scrlfile` authors, IMHO.

**Comment by u-fischer** [comment]
There is also the hopatch package.

**Comment by callegar** [comment]
Right, but this seems to rely on the ability to provide hooks from either `filehook` or `scrlfile` or `memoir`, meaning that if LaTeX ends up providing its own interface for hooks and `filehook` and `scrlfile` are turned into thin wrappers to re-deliver it with compatibility interfaces, `hopatch` may work seamlessly.

Sorry for having been late in talking with the `scrlfile` and `filehook` authors. Had a few problems lately, but I'll try to catch up.

**Comment by komascript** [comment]
I would either need three hooks and the possibility to change the `<file-name>` argument of `\InputIfFileExists` or one hook in `\IfFileExists` and two hooks in `\InputIfFileExists` and (if possible) usage of `\@addtofilelist{\@curr@file}` instead of `\@addtofilelist{#1}` in `\InputIfFileExists`, e.g., (this is only a first suggestion, I've done some tests, but I'm far away of doing deep inspections of code and haven't done a lot of tests):

```LaTeX
% Change of \IfFileExists to have a hook after \@curr@file has been build.
% Hook code can operate on \@curr@file and even could change the file name.
\DeclareRobustCommand\IfFileExists[1]{%
  \set@curr@file{#1}%
  \csname IfFileExists@early@h@@k\endcsname% MJK: added
  \expandafter\IfFileExists@\expandafter{\@curr@file}}
% Add hooks before and after \@@input of the file.
% The hook before could use \@curr@file instead of #1 but \@@input
% can change \@curr@file, so it wouldn't be able to use it. Maybe something
% like this would work. The \@swaptwoargs is not longer needed, because
% \@filef@und is completely expanded before #2 and the input. A significant
% change it, that not #1 but \@curr@file is added to the file list. I think
% this should be.
\DeclareRobustCommand \InputIfFileExists[2]{%
  \IfFileExists{#1}{%
    \csname InputIfFileExists@early@h@@k\endcsname
    \edef\reserved@a{%
      \unexpanded{#2}\noexpand\@addtofilelist{\@curr@file}\noexpand\@@input%
      \@filef@und
      \noexpand\def\noexpand\@curr@file{\@curr@file}% restore \@curr@file after loading the found file
    }%
    \reserved@a
    \csname InputIfFileExists@late@h@@k\endcsname
  }%
}
```

With this or a similar code in LaTeX, I would not need to use my own definition of `\InputIfFileExists` in `scrlfile` but simply can use something like:

```LaTeX
\providecommand*{\InputIfFileExists@early@h@@k}{}
\g@addto@macro\InputIfFileExists@early@h@@k{%
  \scr@load@hook{before}{\@curr@file}%
}
\providecommand*{\InputIfFileExists@late@h@@k}{}
\g@addto@macro\InputIfFileExists@late@h@@k{%
  \scr@load@hook{after}{\@curr@file}%
}
\providecommand*{\IfFileExists@early@h@@k}{}
\g@addto@macro\IfFileExists@early@h@@k{%
  \scr@recursive@alias@walk
}
\newcommand*{\scr@recursive@alias@walk}{%
  \expandafter\ifx\csname \@curr@file-@alias\endcsname\relax\else
    \expandafter\let\expandafter\@curr@file
    \csname \@curr@file-@alias\endcsname
    \expandafter\scr@recursive@alias@walk
  \fi
}
```

I still would need to patch `\@onefilewithoptions` and `@loadfilewithoptions` but these are not completely new definitions but wrappers around the original definitions. So changing them in the LaTeX kernel or by other packages (which are loaded before `scrlfile`) mostly doesn't matter.

@u-fischer hopatch does not implements own hooks but only provides a general interface around the hooks of other packages. So with a new kernel it could either be used unchanged or instead also directly use the generic interface of the LaTeX kernel.

**Comment by FrankMittelbach** [comment]
By end of this week we will have a new dev release on CTAN  what will provide a general hook management and also provide a larger set of hooks into the file loading area. This will then us give 2-3 months where we can make sure that provided hooks are sufficient for that space and have adjustments made in other packages to use the hooks. 

Packages like filehook, scrlfile etc can then as a first step offer their existing interfaces by internally using the kernel hooks rather than patching int the kernel commands and as a second step gradually fading out the  packages in other places (and use the kernel hooks there directly).

If it turns out that a necessary hook is still missing then the next 2 months should allow us to correct that.

We plan to push that dev release probably tomorrow.

**Comment by FrankMittelbach** [comment]
I should have said the the intention here is that once this is in place there is no need to patch *any* of the internals but rely only on using the hook management to add code to the kernel commands!  Thus if it is determined that KOMA, say, still needs to patch rather then just use the hooks then this needs further analysis and appropriate action/correction.

**Comment by callegar** [comment]
This looks like a very good plan. Maybe we should get in touch explicitly with the authors of KOMA (`scrlfile) and `filehook` to make sure that their comments can be collected early, in case they miss the message here (e.g. because of the time of the year when they might be on vacation). Would you like me to do that?

**Comment by FrankMittelbach** [comment]
@callegar we have this on our list, but I thought we wait with it until the dev release is available so that they don't have to grab it from github but rather simply use it on miktex or texlive.

**Comment by FrankMittelbach** [comment]
@callegar there are in fact more people to inform (you included, if I remember correctly because of everypage or some such package) but if you want to take on that communication, once the code and documentation is easily available, it would certainly take off one action item from our list.

**Comment by komascript** [comment]
@FrankMittelbach The new hooks are very nice, but IMHO they do not allow to change the input file name and therefore are not enough to implement `\ReplaceInput`, `\ReplaceClass`, `\ReplacePackage` of `scrlfile`. So I would still need to patch `\InputIfFileExists` or `\IfFileExists` to implement them. See https://github.com/latex3/latex2e/issues/225#issuecomment-581132714 above.

To see whether and how it is possible to avoid patching `\RequirePackage` to realise `\PreventPackageFromLoading` I need more time (preferably after vacation). And so for all the variants of `\AfterPackage` etc. Most of them are made because of user acclamations and lack of comprehensive test cases. I'm using `\AfterPackage` and `\AfterPackage*` only. And the implementation of both is very easy with the new hooks.

**Comment by FrankMittelbach** [comment]
> @FrankMittelbach The new hooks are very nice, but IMHO they do not allow to change the input file name and therefore are not enough to implement `\ReplaceInput`, `\ReplaceClass`, `\ReplacePackage` of `scrlfile`. So I would still need to patch `\InputIfFileExists` or `\IfFileExists` to implement them. See [#225 (comment)](https://github.com/latex3/latex2e/issues/225#issuecomment-581132714) above.

Hi Markus, there are even possibilities to do this right now with the current hooks (as long the first line of the package contains no code) but that may not  not be the best approach, so we are looking at this already.

On the whole I must confess, however, that I'm not sure that LaTeX (or a package) should provide this functionalty as a general user-level option. As an exception of some sort it might be useful but if you think that packages or classes start all kind replacements of that nature and that gets cascaded ...

Basically what I'm trying to say is that this needs a good and reliable concept before we should offer that from a kernel level.

> To see whether and how it is possible to avoid patching `\RequirePackage` to realise `\PreventPackageFromLoading` I need more time (preferably after vacation). And so for all the variants of `\AfterPackage` etc. Most of them are made because of user acclamations and lack of comprehensive test cases. I'm using `\AfterPackage` and `\AfterPackage*` only. And the implementation of both is very easy with the new hooks.

As I said, something like

     \AddToHook{package/before/foo}{\RequirePackage{foo}\endinput}

does the trick already as long as the first line of foo doesn't contain code but this should probably not the last answer on this.

**Comment by u-fischer** [comment]
> Falk Hanisch uses `\PreventPackageFromLoading` a lot in `tudscr`.

personally I don't like this neither. It would be saner to make a long list of unwanted packages, to document this, to check at begin document and then to error, so that the user knows that the package should not be used. It is more work for the user, but then at least they know what's going on. Doing it more or less silently makes debugging such classes/templates quite a nightmare.

**Comment by komascript** [comment]
> Doing it more or less silently makes debugging such classes/templates quite a nightmare.

It's not silently. Falk uses the optional argument of `\PreventPackageFromLoading` at least to warn the user. And it may be an argument to say: "If we prevent loading the package, we do not need to test, if the user has ignored the advise not to do so." But I cannot speak for him. I've used his CI classes only as an example that the command is already used and that I cannot simply remove it from `scrlfile` without breaking existing classes/packages/documents.

**Comment by FrankMittelbach** [comment]
Well, as I said, one could right now do something. I haven't checked how the definition really is (I think it supports a clist of packages) but something like
```
\newcommand\PreventPackageFromLoading[2][]
   {%
     \AddToHook{package/before/#2}{\RequirePackage{#2}\endinput}%
     #1
   }

\PreventPackageFromLoading[\typeout{--- Package trace ignored ---}]{trace}   
```
would work for now. 

As to the \Replace... that is a bit more tricky as you can't rely on the first line of an arbitrary file being harmless.

**Comment by blefloch** [comment]
Your code seems wrong, Frank, maybe you mixed up `\PreventPackageFromLoading` and `\Replace...`.  For the former, remove `\RequirePackage{#2}` I suppose.

**Comment by FrankMittelbach** [comment]
Don't think so Bruno. The code requires the package and then cancels the load after the first line (which is why this line better is a comment or somehting harmless (which is usually the case with packages)). Without `\RequirePackage` the `\endinput` would stop reading from the *current* file. Anyway, I don't think this is a long term solution but it is perhaps good enough for a while to implement such a comment without needing to resort to patching into various kernel commands again.

A replace command could work similarly, ie first require and cancel one and then load the other file but for general files that would be rather dangerious as the first line might contain anything.

**Comment by blefloch** [comment]
You're probably right, Frank but could you explain when the `package/before/#2` hook is run?  I thought this would be triggered by the user doing `\usepackage{#2}`, then the hook would be run, and doing `\endinput` would be the right thing.

**Comment by mrpiggi** [comment]
> > Falk Hanisch uses `\PreventPackageFromLoading` a lot in `tudscr`.
> 
> personally I don't like this neither. It would be saner to make a long list of unwanted packages, to document this, to check at begin document and then to error, so that the user knows that the package should not be used. 

I use `\PreventPackageFromLoading` to ensure a high degree of compatibility with other classes or packages. For example, `tudscr` uses packages `geometry` and `scrlayer-scrpage` by default, which leads to problems in conjunction with the `standalone` class. Just try this:
```latex
\documentclass[crop]{standalone}
\usepackage{geometry}
\begin{document}
\rule{50pt}{50pt}
\end{document}
```
And since the packages are loaded by `tudscr`, the user can do very little about it. So I do not always prevent loading, but only under certain circumstances. Another use case is to delay the loading of a package to a later time.  I would not like to give up the functionality, that would be a pity.

**Comment by FrankMittelbach** [comment]
> You're probably right, Frank but could you explain when the `package/before/#2` hook is run? I thought this would be triggered by the user doing `\usepackage{#2}`, then the hook would be run, and doing `\endinput` would be the right thing.

No I'm not right, you are  --- as usual, the \RequirePackage is not needed in this case.
```
\documentclass{article}

\newcommand\PreventPackageFromLoading[2][]
   {\AddToHook{package/before/#2}{\endinput}#1}

\newcommand\ReplacePackage[3][]
   {\AddToHook{package/before/#2}{\RequirePackage{#3}\relax\endinput}#1}

\PreventPackageFromLoading[\typeout{--- Package url prevent from laoding ---}]{url}
\ReplacePackage[\typeout{--- Package trace replaced by varioref ---}]{trace}{varioref}   

\usepackage{url}
\usepackage{trace}

\show\url  \show\traceon  \show\vref
\stop
```

**Comment by FrankMittelbach** [comment]
> Another use case is to delay the loading of a package to a later time. I would not like to give up the functionality, that would be a pity.

@mrpiggi One of the reasons to introduce the hook management is that the need for ordering package loading should become less important and over time hopefully vanish. But with the above suggested implementation that use case would not be supported, because the package is loaded (only very briefly and only its first line), but without further adjustments that would top it from being loaded later.

**Comment by u-fischer** [comment]
> > personally I don't like this neither. It would be saner to make a long list of unwanted packages, to document this, to check at begin document and then to error, so that the user knows that the package should not be used.
> 
> I use `\PreventPackageFromLoading` to ensure a high degree of compatibility with other classes or packages. For example, `tudscr` uses packages `geometry` and `scrlayer-scrpage` by default, which leads to problems in conjunction with the `standalone` class. Just try this:
> 
> ```
> \documentclass[crop]{standalone}
> \usepackage{geometry}
> \begin{document}
> \rule{50pt}{50pt}
> \end{document}
> ```
> 

You don't need to prevent geometry from loading. Simply `\PassOptionsToPackage{pass}{geometry}` would be okay too. Or tell your user not to load geometry (as it is unneeded) and error if it is there with standalone. 

And regarding compability: If you want to do everything in your own way it would be much better to do it like memoir does it: by emulating packages like you want them to be, instead of loading them and then to patch and redefine their core commands. Then you have everything under control and if there is a bug somewhere, it doesn't require long investigations if the original package or your patches are the problem.

**Comment by komascript** [comment]
> And regarding compability: If you want to do everything in your own way it would be much better to do it like memoir does it: by emulating packages like you want them to be, instead of loading them and then to patch and redefine their core commands. Then you have everything under control and if there is a bug somewhere, it doesn't require long investigations if the original package or your patches are the problem.

Doing so was part of the `\InputIfFileExists` problem, because copying parts of `scrlfile` into `memoir` made `memoir` incompatible with `scrlfile` and resulted in problems, when users used `scrlfile` directly or indirectly. I have had to change `scrlfile` to make it compatible with `memoir` again. So I cannot recommend this way.

And with this suggestion in last consequence I would have to make my own format.

I currently have an idea how to make `\PreventPackageFromLoading` (and `\UnpreventPackageFromLoading`) with the new hooks. I will think about it and try it out after summer holiyday in BW (maybe earlier depending on weather conditions).

**Comment by mrpiggi** [comment]
> You don't need to prevent geometry from loading. Simply `\PassOptionsToPackage{pass}{geometry}` would be okay too. 

In this particular case, something like
```latex
\AfterClass*{standalone}{\PassOptionsToPackage{pass}{geometry}}
```
could work in order to use
```latex
\documentclass[class=tudscr...]{standalone}
```
but the issue with `scrlayer-scrpage` would still exist, although I maybe could use `\AfterClass` in this case as well. Nevertheless, delaying the moment of loading a package is still an issue.

> Or tell your user not to load geometry (as it is unneeded) and error if it is there with standalone.

As I said, both named packages are loaded in `tudscr` by default.

> And regarding compability: If you want to do everything in your own way it would be much better to do it like memoir does it: by emulating packages like you want them to be, instead of loading them and then to patch and redefine their core commands.
 
In principle, you're right. On the other hand, it is a huge effort to emulate very large packages with thousands lines of code, just to make small changes and to constantly adapt to new features and bug fixes after an update.

**Comment by daleif** [comment]
My development version of memoir will now look for the format >= 2020/10/01 format, and if true it will use the new file hooks to implement \AtBegin/EndFile/Package/Class. Then the kernel version of \InputIfFileExists will no be touched by memoir.

The file hooks will look like this

```
\IfFormatAtLeastTF{2020/10/01}{
  \newcommand\AtBeginFile[2]{\AddToHook{file/before/#1}{#2}}
  \newcommand\AtEndFile[2]{\AddToHook{file/after/#1}{#2}}

  \newcommand\AtBeginPackage[2]{\AddToHook{package/before/#1}{#2}}
  \newcommand\AtEndPackage[2]{\AddToHook{package/after/#1}{#2}}
  \newcommand{\RequireAtEndPackage}[2]{%
    \@ifpackageloaded{#1}{#2}{\AtEndPackage{#1}{#2}}}

  \newcommand\AtBeginClass[2]{\AddToHook{class/before/#1}{#2}}
  \newcommand\AtEndClass[2]{\AddToHook{class/after/#1}{#2}}
  \newcommand{\RequireAtEndClass}[2]{%
    \@ifclassloaded{#1}{#2}{\AtEndClass{#1}{#2}}}

}{
   % old nasty code
}
```

As these are user commands I haven't added hook labels to them

I'm also going to change how trimmarks are added, which now will be done via shipout/background, this hook will be labelled as memoir/trimmarks as it is likely that some users may want to add stuff before it. A poor mans page color test works just fine with the dev format.

**Comment by FrankMittelbach** [comment]
Looks good and not setting hook labels is the correct thing here (it is in most circumstances anyway as then the label is build based on context).
